salt/pillar/nsot.py

```diff
@@ -0,0 +1,181 @@
+    with this character replacing dots. For example, with this option set to
+     '-', the minion rtr1.lax1 would be looked up in NSoT as rtr1-lax1.
+
+all_device_info: ``list``
+    This is a list of minions that will receive all device
+    information in NSoT. For example, if this list contained the entry
+    'test_min', then the minion 'test_min' would have all devices in NSoT under
+    pillar['nsot']['devices']
+
+minion_regex: ``list``
+    This is in case you only want certain minions being looked
+    up in NSoT. For example, if you have 500 minions on a single master, you may
+    not want all of them hitting the NSoT endpoint. With this list of regexes,
+    only minions matching the regex will be queried.
+
+Here's an example config with all options:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - nsot:
+        api_url: http://nsot_url.com/api/
+        email: 'user@site.com'
+        secret_key: abc234
+        replace_dot: '-'
+        all_device_info:
+          - minion_1
+          - minion_2
+        minion_regex:
+          - 'rtr*'
+          - 'sw*'
+          - '^router*'
+
+'''
+
+from __future__ import absolute_import, print_function, unicode_literals
+import logging
+import re
+
+# Import Salt libs
+import salt.utils.http
+
+log = logging.getLogger(__name__)
+
+
+def _get_token(url, email, secret_key):
+    '''
+    retrieve the auth_token from nsot
+
+    :param url: str
+    :param email: str
+    :param secret_key: str
+    :return: str
+    '''
+    url += 'authenticate/'
+    data_dict = {"email": email, "secret_key": secret_key}
+    query = salt.utils.http.query(url, data=data_dict, method='POST',
+                                  decode=True)
+    error = query.get('error')
+    if error:
+        log.error('can\'t get auth_token from nsot! reason: %s', error)
+        return False
+    else:
+        log.debug('successfully obtained token from nsot!')
+        return query['dict']['auth_token']
+
+
+def _check_minion_regex(minion_id, minion_regex):
+    '''
+    check whether or not this minion should have this external pillar returned
+
+    :param minion_id: str
+    :param minion_regex: list
+    :return: bool
+    '''
+    get_pillar = False
+    for pattern in minion_regex:
+        log.debug('searching %s using %s', minion_id, minion_regex)
+        match = re.search(pattern, minion_id)
+        if match and match.string == minion_id:
+            log.debug('found match! %s: regex %s', minion_id, minion_regex)
+            get_pillar = True
+            break
+        log.debug('unable to match %s using regex %s', minion_id, minion_regex)
+    return get_pillar
+
+
+def _query_nsot(url, headers, device=None):
+    url += 'devices/'
+    if not device:
+        query = salt.utils.http.query(url, header_dict=headers, decode=True)
+    else:
+        device += '/'
+        query = salt.utils.http.query(url + device, header_dict=headers,
+                                      decode=True)
```

Comment:
Please refer to a proper way of doing this, i.e. `os.path.join`, instead of `+=` on strings, and then refactor accordingly.